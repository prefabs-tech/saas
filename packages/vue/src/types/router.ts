import type { RouteComponent } from "vue-router";

interface RouteOverride {
  component?: RouteComponent;
  meta?: {
    layout?: RouteComponent;
  };
  path?: string;
}

interface RouteOverrides {
  acceptInvitation: RouteOverride;
  changePassword: RouteOverride;
  google?: RouteOverride;
  home?: string;
  login: RouteOverride;
  signup: RouteOverride & { disabled?: boolean };
  signupFirstUser: RouteOverride & { disabled?: boolean };
  passwordReset: RouteOverride;
  passwordResetRequest: RouteOverride;
  passwordResetRequestAcknowledge: RouteOverride;
  profile?: RouteOverride;
  roles?: RouteOverride;
  verifyEmail?: RouteOverride;
  verifyEmailReminder?: RouteOverride;
}

export type { RouteOverride, RouteOverrides };
