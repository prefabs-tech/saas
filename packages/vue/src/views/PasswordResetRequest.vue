<template>
  <Page
    :title="t('user.passwordResetRequest.title')"
    class="auth password-reset-request"
  >
    <slot name="instructions"></slot>

    <PasswordResetRequestForm @submit="handleSubmit" />
  </Page>
</template>

<script lang="ts">
export default {
  name: "PasswordResetRequest",
};
</script>

<script setup lang="ts">
import { useI18n } from "@dzangolab/vue3-i18n";
import { Page } from "@dzangolab/vue3-ui";
import { useRouter } from "vue-router";

import PasswordResetRequestForm from "../components/PasswordResetRequestForm.vue";
import { useTranslations } from "../index";
import useUserStore from "../store";

import type { PasswordResetRequestPayload } from "../types";

const messages = useTranslations();

const { t } = useI18n({ messages });

const router = useRouter();

const { requestPasswordReset } = useUserStore();

const handleSubmit = async (payload: PasswordResetRequestPayload) => {
  await requestPasswordReset(payload).then(() => {
    router.push({
      name: "resetPasswordRequestAcknowledge",
      query: { email: payload.email },
    });
  });
};
</script>
