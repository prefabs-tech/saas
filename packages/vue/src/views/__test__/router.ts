import { createRouter, createWebHistory } from "vue-router";

import Accounts from "../accounts/Index.vue";

const Test = {
  name: "Test",
  template: `<h1>Test</h1>`,
};

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      component: Accounts,
      name: "accounts",
      path: "/accounts",
    },
    {
      component: Test,
      path: "/",
    },
  ],
});

export default router;
