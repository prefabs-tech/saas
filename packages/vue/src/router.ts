import { storeToRefs } from "pinia";
import { Router } from "vue-router";

import AuthGoogleCallback from "./components/AuthGoogleCallback.vue";
import useUserStore from "./store";
import { getVerificationStatus } from "./supertokens";
import AcceptInvitation from "./views/AcceptInvitation.vue";
import ChangePassword from "./views/ChangePassword.vue";
import VerifyEmailReminder from "./views/EmailVerificationReminder.vue";
import Login from "./views/Login.vue";
import PasswordReset from "./views/PasswordReset.vue";
import PasswordResetRequest from "./views/PasswordResetRequest.vue";
import PasswordResetRequestAcknowledge from "./views/PasswordResetRequestAcknowledge.vue";
import Profile from "./views/Profile.vue";
import Roles from "./views/Roles/Index.vue";
import Signup from "./views/Signup.vue";
import SignupFirstUser from "./views/SignupFirstUser.vue";
import VerifyEmail from "./views/VerifyEmail.vue";

import type {
  DzangolabVueUserConfig,
  RouteOverride,
  RouteOverrides,
} from "./types";
import type { RouteMeta, RouteRecordRaw } from "vue-router";

const _routes = {
  acceptInvitation: {
    component: AcceptInvitation,
    name: "acceptInvitation",
    path: "/signup/token/:token?",
  } as RouteRecordRaw,
  changePassword: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: ChangePassword,
    name: "changePassword",
    path: "/change-password",
  },
  google: {
    component: AuthGoogleCallback,
    name: "authcallback",
    path: "/auth/callback/google",
  },
  login: {
    component: Login,
    name: "login",
    path: "/login",
  } as RouteRecordRaw,
  profile: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: Profile,
    name: "profile",
    path: "/profile",
  } as RouteRecordRaw,
  roles: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: Roles,
    name: "roles",
    path: "/roles",
  } as RouteRecordRaw,
  signup: {
    component: Signup,
    name: "signup",
    path: "/signup",
  } as RouteRecordRaw,
  signupFirstUser: {
    component: SignupFirstUser,
    name: "signupFirstUser",
    path: "/signup-first-user",
  } as RouteRecordRaw,
  passwordReset: {
    component: PasswordReset,
    name: "resetPassword",
    path: "/reset-password/:token?",
  } as RouteRecordRaw,
  passwordResetRequest: {
    component: PasswordResetRequest,
    name: "resetPasswordRequest",
    path: "/reset-password-request",
  } as RouteRecordRaw,
  passwordResetRequestAcknowledge: {
    component: PasswordResetRequestAcknowledge,
    name: "resetPasswordRequestAcknowledge",
    path: "/reset-password-request-acknowledge",
  },
  verifyEmail: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: VerifyEmail,
    name: "verifyEmail",
    path: "/verify-email",
  },
  verifyEmailReminder: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: VerifyEmailReminder,
    name: "verifyEmailReminder",
    path: "/verify-email-reminder",
  },
};

const getRoute = (
  defaultRoute: object,
  override?: RouteOverride,
): RouteRecordRaw => {
  return {
    ...defaultRoute,
    ...override,
  } as RouteRecordRaw;
};

const addRoutes = (router: Router, userConfig?: DzangolabVueUserConfig) => {
  const routes: RouteOverrides | undefined = userConfig?.routes;

  router.addRoute(getRoute(_routes.google, routes?.google));

  router.addRoute(getRoute(_routes.login, routes?.login));

  if (!routes?.signup?.disabled) {
    router.addRoute(getRoute(_routes.signup, routes?.signup));
  }

  if (routes?.signup?.disabled && !routes?.signupFirstUser?.disabled) {
    router.addRoute(getRoute(_routes.signupFirstUser, routes?.signupFirstUser));
  }

  router.addRoute(getRoute(_routes.passwordReset, routes?.passwordReset));

  router.addRoute(
    getRoute(_routes.passwordResetRequest, routes?.passwordResetRequest),
  );

  router.addRoute(getRoute(_routes.profile, routes?.profile));

  router.addRoute(getRoute(_routes.roles, routes?.roles));

  router.addRoute(
    getRoute(
      _routes.passwordResetRequestAcknowledge,
      routes?.passwordResetRequestAcknowledge,
    ),
  );

  router.addRoute(
    getRoute(
      _routes.changePassword,
      routes?.changePassword,
    ),
  );

  router.addRoute(getRoute(_routes.acceptInvitation, routes?.acceptInvitation));

  if (userConfig?.features?.signUp?.emailVerification) {
    router.addRoute(getRoute(_routes.verifyEmail, routes?.verifyEmail));

    router.addRoute(
      getRoute(_routes.verifyEmailReminder, routes?.verifyEmailReminder),
    );
  }
};

const redirectRoutes = (router: Router) => {
  router.beforeEach((to, from, next) => {
    const userStore = useUserStore();
    const { user } = storeToRefs(userStore);

    const routesToRedirect = [
      "acceptInvitation",
      "login",
      "signup",
      "signupFirstUser",
      "resetPassword",
      "resetPasswordRequest",
    ];
    const name = to.name as string;

    if (user.value && routesToRedirect.includes(name)) {
      next({ name: "profile" });
    } else {
      next();
    }
  });
};

const addAuthenticationGuard = (
  router: Router,
  userConfig?: DzangolabVueUserConfig,
) => {
  router.beforeEach(async (to) => {
    const meta = to.meta as RouteMeta;

    const userStore = useUserStore();

    const EmailVerificationEnabled =
      !!userConfig?.features?.signUp?.emailVerification;

    const name = to.name as string;
    const routesToRedirect = ["verifyEmail", "verifyEmailReminder"];
    const { user } = storeToRefs(userStore);
    const { getUser } = userStore;

    if (!user.value) {
      user.value = await getUser();
    }

    if (meta.authenticated && !user.value) {
      router.push({ name: "login" }); // using next inside async function is not allowed
    }

    if (EmailVerificationEnabled && user.value) {
      const isEmailVerified = await getVerificationStatus();

      if (
        meta.authenticated &&
        !isEmailVerified &&
        !routesToRedirect.includes(name)
      ) {
        router.push({ name: "verifyEmailReminder" });
      } else if (isEmailVerified && routesToRedirect.includes(name)) {
        router.push({ name: "home" });
      }
    }

    const isProfileCompleted = !!user.value?.isProfileCompleted;
    const profileCompletionEnabled = user.value?.isProfileCompleted !== undefined;

    if (meta.authenticated && profileCompletionEnabled && !isProfileCompleted && name !== "profile") {
      router.push({ name: "profile" });
    }
  });
};

const updateRouter = (router: Router, userConfig?: DzangolabVueUserConfig) => {
  addRoutes(router, userConfig);

  addAuthenticationGuard(router, userConfig);

  redirectRoutes(router);
};

export default updateRouter;
