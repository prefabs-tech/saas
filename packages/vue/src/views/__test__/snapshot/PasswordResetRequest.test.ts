import i18Plugin, { useLocaleStore } from "@dzangolab/vue3-i18n";
import { LoadingButton, LoadingIcon } from "@dzangolab/vue3-ui";
import { mount, RouterLinkStub } from "@vue/test-utils";
import { createPinia } from "pinia";
import { describe, it, expect } from "vitest";

import appConfig from "../../../components/__test__/config";
import PasswordResetRequest from "../../PasswordResetRequest.vue";
import router from "../router";

import type { VueWrapper } from "@vue/test-utils";

describe("PasswordResetRequest", () => {
  const pinia = createPinia();

  const { setLocale } = useLocaleStore(appConfig.slug);

  const locales = ["en", "fr"];

  for (const locale of locales) {
    setLocale(locale);

    const wrapper: VueWrapper = mount(PasswordResetRequest, {
      global: {
        components: {
          LoadingButton,
          LoadingIcon,
        },
        plugins: [
          pinia,
          [
            i18Plugin,
            {
              config: appConfig,
            },
          ],
          router,
        ],
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    it("matches snapshot in " + locale, () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  }
});
