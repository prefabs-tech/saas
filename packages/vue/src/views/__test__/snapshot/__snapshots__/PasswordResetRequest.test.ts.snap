// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PasswordResetRequest > matches snapshot in en 1`] = `
<div
  class="page auth password-reset-request"
  data-centered="false"
>
  <div
    class="page-header"
  >
    <div
      class="page-title-wrapper"
    >
      <div
        class="title"
      >
        <h1>
          Reset your password
        </h1>
        
        <!---->
        
      </div>
      
      <!---->
      
    </div>
    <div
      class="page-toolbar"
    >
      
      
    </div>
  </div>
  <div
    class="page-content"
  >
    
    
    
    <form
      novalidate=""
    >
      <div
        class="field email"
      >
        <label
          for="email"
        >
          Your email
        </label>
        
        <input
          class=""
          name="email"
          placeholder="john.doe@example.com"
          tabindex="0"
          type="email"
          value=""
        />
        <!---->
        
      </div>
      <div
        class="actions"
      >
        <button
          class="loading-button"
          data-v-3fdbc7b3=""
        >
          <div
            class="label"
            data-v-3fdbc7b3=""
          >
            <span
              data-v-3fdbc7b3=""
            >
              Email me
            </span>
          </div>
          <!---->
        </button>
      </div>
    </form>
    
    <!---->
  </div>
</div>
`;

exports[`PasswordResetRequest > matches snapshot in fr 1`] = `
<div
  class="page auth password-reset-request"
  data-centered="false"
>
  <div
    class="page-header"
  >
    <div
      class="page-title-wrapper"
    >
      <div
        class="title"
      >
        <h1>
          Réinitialisation de votre mot de passe
        </h1>
        
        <!---->
        
      </div>
      
      <!---->
      
    </div>
    <div
      class="page-toolbar"
    >
      
      
    </div>
  </div>
  <div
    class="page-content"
  >
    
    
    
    <form
      novalidate=""
    >
      <div
        class="field email"
      >
        <label
          for="email"
        >
          Votre adresse courriel
        </label>
        
        <input
          class=""
          name="email"
          placeholder="john.doe@example.com"
          tabindex="0"
          type="email"
          value=""
        />
        <!---->
        
      </div>
      <div
        class="actions"
      >
        <button
          class="loading-button"
          data-v-3fdbc7b3=""
        >
          <div
            class="label"
            data-v-3fdbc7b3=""
          >
            <span
              data-v-3fdbc7b3=""
            >
              Réinitialiser
            </span>
          </div>
          <!---->
        </button>
      </div>
    </form>
    
    <!---->
  </div>
</div>
`;
